updates for group management concept page (#3163)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/scim/groups.mdx

The following table provides more information about how SCIM groups relate to HCP Terraform teams.

| | SCIM groups | HCP Terraform teams |
| --- | --- | --- |
| Description | Representations of IdP groups that automatically create corresponding teams in HCP Terraform. | Organization-scoped entities that control access to workspaces, projects, and other resources. Teams have associated permissions and can contain members. |
| Management | Managed by your IdP through SCIM API calls. | Automatically created and managed by SCIM based on IdP groups. |
| Scope | Organization-scoped in HCP Terraform. | Scoped to a specific organization in HCP Terraform. |
| Permissions | No permissions. | Configurable permissions for workspaces, projects, and organization-level access. |
| Membership source | IdP. | Automatically synchronized from the corresponding IdP group. |
| Team creation | Provisioning a group automatically creates a team. | Created automatically when the corresponding SCIM group is provisioned. |

When a group is provisioned, HCP Terraform performs the following actions:

> 1. Creates a team with the same name as the IdP group's `displayName`.

Reviewer: Not sure exactly where to put this, but we should mention somewhere that if a team exists with the same name already, then we link the IdP group to that existing team and synchronize the membership.

Author: Added a new section to explain behavior around existing teams.

1. Synchronizes the group's membership to the team.

> 1. Adds users to the organization if they are not already members.

Reviewer: ```suggestion
```

This isn't done in TFC. Users are added as members of the organization as part of the user provisioning.

<!-- END: TFEnterprise:only name:group-to-team-links -->
## Group lifecycle

SCIM groups follow a standard create, update, and delete lifecycle managed by your identity provider.

### Create stage

<!-- BEGIN: TFEnterprise:only name:tfe-create -->
Terraform Enterprise performs the following actions when your IdP creates a group through SCIM:

1. Generates a unique SCIM `id` for the group.
1. Stores the group with its `displayName` and any `externalId` from the IdP.

Terraform Enterprise teams aren't linked to the group until you explicitly create a mapping to establish the connection. After creation, the group is available for team mapping.
<!-- END: TFEnterprise:only name:tfe-create -->

<!-- BEGIN: TFC:only name:tfc-create -->
HCP Terraform performs the following actions when your IdP creates a group through SCIM:

1. Generates a unique SCIM `id` for the group.
1. Stores the group with its `displayName` and any `externalId` from the IdP.
1. Automatically creates a team with the same name as the group's `displayName`.
1. Synchronizes the group's membership to the newly created team.
<!-- END: TFC:only name:tfc-create -->

> SCIM group `displayName` values must be unique across <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product -->. Values aren't case-sensitive. As a result, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> considers `Engineering` and `engineering` to be the same `displayName`. When you attempt to provision a group with an existing `displayName`, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> rejects the request with HTTP `409 Conflict`.

Reviewer: This might be better off in the create SCIM group API docs, but I was thinking should call out that `owners` and `sso` are reserved names that cannot be used as SCIM group names.

Author: part of new section

> When <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> adds groups, it preserves the original letter case of SCIM group display names when it stores them. API responses return the stored `displayName` with that preserved letter case.

Reviewer: We may want to also mention here that if a SCIM group is linked to an existing team and the case doesn't match, we will update the existing team's name to match the case of the SCIM group.

Author: part of new section

> To reference a group member in the `members[].value` object, you must include the user's public SCIM `id` value from `/scim/v2/Users`. Before you can perform group membership operations, the user must already be provisioned in Terraform Enterprise through SCIM.

Reviewer: There's a Terraform Enterprise reference here.

### Update stage

Your IdP can update the following group properties:

> - Group display name changes. <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> stores the updates uisng the letter case provided by the IdP.

Reviewer: ```suggestion
- Group display name changes. <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> updates the corresponding team name to match the name provided by the IdP.
```

Reviewer: My thought here was to clarify that we will be updating the team name to match the display name, as opposed to TFE where the team names aren't affected.

- **Site admin group**: The group configured as the site administrator group in SCIM settings cannot be mapped to regular Terraform Enterprise teams.
<!-- END: TFEnterprise:only name:group-to-team-links -->
## Group lifecycle

SCIM groups follow a standard create, update, and delete lifecycle managed by your identity provider.

### Create stage

<!-- BEGIN: TFEnterprise:only name:tfe-create -->
Terraform Enterprise performs the following actions when your IdP creates a group through SCIM:

1. Generates a unique SCIM `id` for the group.
1. Stores the group with its `displayName` and any `externalId` from the IdP.

Terraform Enterprise teams aren't linked to the group until you explicitly create a mapping to establish the connection. After creation, the group is available for team mapping.
<!-- END: TFEnterprise:only name:tfe-create -->

<!-- BEGIN: TFC:only name:tfc-create -->
HCP Terraform performs the following actions when your IdP creates a group through SCIM:

1. Generates a unique SCIM `id` for the group.
1. Stores the group with its `displayName` and any `externalId` from the IdP.

> 1. Automatically creates a team with the same name as the group's `displayName`. 

Reviewer 2: If the team already exists, it links instead of creating, right?

Author: yes, this is described earlier. I'll add a link

1. Synchronizes the group's membership to the newly created team.

> <!-- END: TFC:only name:tfc-create -->

Author: ```suggestion

When a team with the same name already exists, then HCP Terraform links the IdP group to the existing team and synchronizes the membership. Refer to [Existing HCP Terraform teams](#existing-hcp-terraform-teams).
<!-- END: TFC:only name:tfc-create -->
```

> SCIM group `displayName` values must be unique across <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product -->. Values aren't case-sensitive. As a result, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> considers `Engineering` and `engineering` to be the same `displayName`. When you attempt to provision a group with an existing `displayName`, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> rejects the request with HTTP `409 Conflict`.

Reviewer 2: ```suggestion
SCIM group `displayName` values must be unique across <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product -->. Values aren't case-sensitive. As a result, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> considers `Engineering` and `engineering` to be the same `displayName`. When you attempt to provision a new group with an existing `displayName`, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> rejects the request with HTTP `409 Conflict`.
```

Not sure if this makes it easier or harder to understand.  But I think the intent is if the admin tries to hand-create a group in TFE/HCPTF with a name that matches a `displayName` group, TFE/HCPTF errors.

I almost want to suggest adding the product name "...provision a new HCP Terraform group..." but...oof.  If you do that (only if it makes things clearer) it may be better to just duplicate/exclude the entire paragraph.

### Update stage

Your IdP can update the following group properties:

- Group display name changes. <!-- BEGIN: TFC:only name:update -->HCP Terraform updates the corresponding team name to match the name provided by the IdP.<!-- END: TFC:only name:update --><!-- BEGIN: TFEnterprise:only name:update -->Terraform Enterprise stores the updates uisng the letter case provided by the IdP.<!-- END: TFEnterprise:only name:update -->
- Membership changes. Adding or removing users from the group triggers membership synchronization to <!-- BEGIN: TFC:only name:product -->the corresponding HCP Terraform team<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->all linked Terraform Enterprise teams<!-- END: TFEnterprise:only name:product -->.

> You can fully replace group or partially update group updates from supported identity providers. Full replacement updates reconcile the group's membership to the roster provided by the IdP. Partial updates support the common group change patterns used by Okta and Microsoft Entra ID, including display name changes, full member replacement, and incremental member add or remove operations.

Reviewer 2: ```suggestion
You can fully replace or partially update groups from supported identity providers. Full replacement updates reconcile the group's membership to the roster provided by the IdP. Partial updates support the common group change patterns used by Okta and Microsoft Entra ID, including display name changes, full member replacement, and incremental member add or remove operations.
```

Not sure if this is 100% accurate or if it reads more clearly.  The original took some working through to parse.

### Synchronization process

When your IdP sends a membership update, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> performs the following actions:

1. Updates the SCIM group's membership records.

<!-- BEGIN: TFEnterprise:only name:tfc-sync-process -->
1. For each linked Terraform Enterprise team where synchronization is not paused, Terraform Enterprise:
<!-- END: TFEnterprise:only name:tfc-sync-process -->

<!-- BEGIN: TFC:only name:tfc-sync-process -->
1. For the corresponding team:
<!-- END: TFC:only name:tfc-sync-process -->

   - Removes users who are no longer in the SCIM group.
   - Adds users newly added to the SCIM group.

>    - Creates organization membership for users if needed.

Reviewer: ```suggestion
```

<!-- END: TFEnterprise:only name:tfe-sync-scope -->

## Group size limits

To ensure reliable synchronization within transaction timeouts, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> enforces the following limits on group membership:

<!-- BEGIN: TFEnterprise:only name:tfc-group-limit -->
| Limit | Value | Behavior |
|-------|-------|----------|
| Maximum members per SCIM group request | 1,000 | Public SCIM group create or update requests that would create or project a group over this limit return HTTP `413 Payload Too Large` |
<!-- END: TFEnterprise:only name:tfc-group-limit -->

<!-- BEGIN: TFC:only name:tfc-group-limit -->
| Limit | Value | Behavior |
|-------|-------|----------|
| Maximum members per SCIM group request | 1,000 | Public SCIM group create or update requests that would create or project a group over this limit return HTTP `413 Payload Too Large` |

> | Maximum IdP groups | 1,000 | Total number of groups that can be synchronized |

Reviewer: No change required now, but as a heads up, we may be increasing this limit before GA.

After deletion, verify the resulting mapping state and team membership for affected teams. Decide whether each affected team should be managed manually in Terraform Enterprise or linked to a different SCIM group.

If the deleted SCIM group was configured as the site admin group, Terraform Enterprise clears the site admin group mapping. Existing site administrator grants are not automatically revoked.
<!-- END: TFEnterprise:only name:tfe-delete -->
<!-- BEGIN: TFC:only name:tfc-delete -->
HCP Terraform performs the following actions when your IdP deletes a group through SCIM:

1. Removes the SCIM group record.
1. Deletes the corresponding team.

When a team is deleted, all team memberships and team permissions are removed. Users remain in the organization unless they are also removed through SCIM user deprovisioning.
<!-- END: TFC:only name:tfc-delete -->

For exact delete semantics and status codes, refer to the [SCIM Groups API](/terraform/cloud-docs/api-docs/scim/groups).

> TODO -- reconcile the API docs

Reviewer 2: Noting there is still a TODO here, likely intentional to keep it for the rollup

<!-- BEGIN: TFEnterprise:only name:tfc-group-limit -->
| Limit | Value | Behavior |
|-------|-------|----------|
| Maximum members per SCIM group request | 1,000 | Public SCIM group create or update requests that would create or project a group over this limit return HTTP `413 Payload Too Large` |
<!-- END: TFEnterprise:only name:tfc-group-limit -->

<!-- BEGIN: TFC:only name:tfc-group-limit -->
| Limit | Value | Behavior |
|-------|-------|----------|
| Maximum members per SCIM group request | 1,000 | Public SCIM group create or update requests that would create or project a group over this limit return HTTP `413 Payload Too Large` |
| Maximum IdP groups | 1,000 | Total number of groups that can be synchronized |
<!-- END: TFC:only name:tfc-group-limit -->

There are also limits on the size of public SCIM and `PATCH` API requests. Refer to the [Public SCIM API](/terraform/cloud-docs/api-docs/scim) and the [SCIM Groups API](/terraform/cloud-docs/api-docs/scim/groups) for the exact request size, operation count, and response details.

> TODO -- Reconcile API details.

Reviewer 2: Noting another TODO just in case.

## Rate limits

SCIM group operations are subject to rate limiting to protect <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> performance.

<!-- BEGIN: TFEnterprise:only name:tfe-rate-limits -->
Public SCIM provisioning requests and admin team-linking operations use different rate limits. Public SCIM provisioning uses shared rate limits across the public SCIM endpoints, while admin team-linking and pause-toggle operations use separate admin rate limits.

For the exact rate limit values, refer to the [Public SCIM API](/terraform/enterprise/api-docs/scim) and the [Team SCIM Group Mapping API](/terraform/enterprise/api-docs/admin/team-scim-group-mapping).
<!-- END: TFEnterprise:only name:tfe-rate-limits -->

<!-- BEGIN: TFC:only name:tfc-rate-limits -->
Public SCIM provisioning uses shared rate limits across the public SCIM endpoints. For the exact rate limit values, refer to the [Public SCIM API](/terraform/cloud-docs/api-docs/scim).
<!-- END: TFC:only name:tfc-rate-limits -->

## API reference

> TODO -- Update this when we have move clarity on the API docs structure

Reviewer 2: Pretty sure the TODOs are left intentionally, but just in case :)

### IdP-specific behavior

Although different identity providers send membership updates in different formats, <!-- BEGIN: TFC:only name:product -->HCP Terraform<!-- END: TFC:only name:product --><!-- BEGIN: TFEnterprise:only name:product -->Terraform Enterprise<!-- END: TFEnterprise:only name:product --> handles both formats and applies the appropriate membership changes.

> For example, Okta sends `PUT` requests with a complete list of all group members, whereas Microsoft Entra ID sends `PATCH` requests with lists of members to add and remove.

Reviewer: ```suggestion
For example, Microsoft Entra ID sends `PATCH` requests with lists of members to add and remove, whereas Okta can send either `PUT` requests with a complete list of all group members or a `PATCH` requests.
```

Doesn't have to be this exact wording, but Okta sending PUT or PATCH is dependent on how the integration to HCPT/TFE is configured.
